File: diagflow/simulated/cluster.py
"""Simulated cluster environment — provides mock data for demo without real infra."""

from __future__ import annotations

from .scenarios import ALL_SCENARIOS


class SimulatedCluster:
    """A mock cluster that serves pre-generated scenario data.

    In production, this would be replaced by calls to real infrastructure.
    For the demo, it returns realistic-looking logs, configs, and metrics.
    """

    def __init__(self, scenario_name: str = "flink_oom"):
        scenario_fn = ALL_SCENARIOS.get(scenario_name)
        if not scenario_fn:
            msg = f"Unknown scenario '{scenario_name}'. Available: {list(ALL_SCENARIOS.keys())}"
            raise ValueError(msg)
        data = scenario_fn()
        self.context = data["context"]
        self.logs = data.get("logs", {})
        self.config = data.get("config", {})
        self.metrics = data.get("metrics", {})
        self.expected_root_cause = data.get("expected_root_cause", "")
        self.scenario_name = scenario_name

    def get_node_log(self, log_path: str, keywords: str = "", max_lines: int = 50) -> str:
        """Return matching lines from a simulated log file."""
        content = self.logs.get(log_path, "")
        if not content:
            return f"[Simulated] Log file '{log_path}' not found on this cluster."

        lines = content.strip().split("\n")
        if keywords:
            kw = keywords.lower()
            lines = [l for l in lines if kw in l.lower()]

        matched = lines[-max_lines:] if len(lines) > max_lines else lines
        return "\n".join(matched) if matched else f"[Simulated] No lines matching '{keywords}' in {log_path}"

    def get_config(self, config_path: str) -> str:
        """Return a config file content as YAML-like text."""
        import yaml
        data = self.config.get(config_path, {})
        if not data:
            return f"[Simulated] Config file '{config_path}' not found."
        return yaml.dump(data, default_flow_style=False)

    def get_metrics(self, metric_names: list[str] | None = None) -> str:
        """Return metrics as key=value pairs."""
        lines = []
        for k, v in self.metrics.items():
            if metric_names and k not in metric_names:
                continue
            lines.append(f"{k}={v}")
        return "\n".join(lines) if lines else "[Simulated] No metrics available."

    def summary(self) -> str:
        """Human-readable cluster summary."""
        c = self.context
        return (
            f"Cluster: {c.get('cluster_id', 'N/A')}\n"
            f"Region: {c.get('region', 'N/A')}\n"
            f"Component: {c.get('component', 'N/A')} v{c.get('version', 'N/A')}\n"
            f"Problem: {c.get('problem_desc', c.get('problem', 'N/A'))}\n"
            f"Detail: {c.get('detail', '')}"
        )

File: internal/simulated/cluster.go
// Package simulated mirrors diagflow/simulated — a mock cluster that serves
// pre-generated scenario data for demo without real infrastructure.
package simulated

import (
	"context"
	"sort"
	"strconv"
	"strings"

	"github.com/Admaing/diagflow/internal/cluster"
)

// Scenario is the data a scenario function returns.
type Scenario struct {
	Context            map[string]any
	Logs               map[string]string
	Config             map[string]map[string]any
	Metrics            map[string]any
	ExpectedRootCause  string
	ExpectedConfidence string
	ScenarioName       string
}

// Cluster implements cluster.Cluster over a Scenario.
type Cluster struct {
	Scenario *Scenario
}

// Verify compile-time interface satisfaction.
var _ cluster.Cluster = (*Cluster)(nil)

// EnsureNodeData is a no-op for simulated clusters.
func (c *Cluster) EnsureNodeData(ctx context.Context) error { return nil }

// FindNode resolves a node ref (always returns a stub entry for the demo).
func (c *Cluster) FindNode(ref string) map[string]any {
	return map[string]any{"node_name": ref, "node_role": ref}
}

// GetNodeLog reads matching lines from a simulated log file.
func (c *Cluster) GetNodeLog(ctx context.Context, logPath, keywords string, maxLines int) (string, error) {
	content, ok := c.Scenario.Logs[logPath]
	if !ok {
		return "[Simulated] Log file '" + logPath + "' not found on this cluster.", nil
	}
	lines := strings.Split(strings.TrimSpace(content), "\n")
	if keywords != "" {
		kw := strings.ToLower(keywords)
		var kept []string
		for _, l := range lines {
			if strings.Contains(strings.ToLower(l), kw) {
				kept = append(kept, l)
			}
		}
		lines = kept
	}
	if len(lines) > maxLines {
		lines = lines[len(lines)-maxLines:]
	}
	if len(lines) == 0 {
		return "[Simulated] No lines matching '" + keywords + "' in " + logPath, nil
	}
	return strings.Join(lines, "\n"), nil
}

// GetConfig returns a config file content as key=value lines.
func (c *Cluster) GetConfig(ctx context.Context, configPath string) (string, error) {
	data, ok := c.Scenario.Config[configPath]
	if !ok || len(data) == 0 {
		return "[Simulated] Config file '" + configPath + "' not found.", nil
	}
	var keys []string
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var b strings.Builder
	for _, k := range keys {
		b.WriteString(k)
		b.WriteString("=")
		b.WriteString(toString(data[k]))
		b.WriteString("\n")
	}
	return b.String(), nil
}

// GetMetrics returns metrics as key=value lines (optionally filtered).
func (c *Cluster) GetMetrics(ctx context.Context, metricNames []string) (string, error) {
	var keys []string
	for k := range c.Scenario.Metrics {
		if len(metricNames) > 0 && !contains(metricNames, k) {
			continue
		}
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var b strings.Builder
	for _, k := range keys {
		b.WriteString(k)
		b.WriteString("=")
		b.WriteString(toString(c.Scenario.Metrics[k]))
		b.WriteString("\n")
	}
	if len(keys) == 0 {
		return "[Simulated] No metrics available.", nil
	}
	return b.String(), nil
}

// Context returns the scenario context.
func (c *Cluster) Context() map[string]any { return c.Scenario.Context }

// ExpectedRootCause returns the scenario's expected root cause.
func (c *Cluster) ExpectedRootCause() string { return c.Scenario.ExpectedRootCause }

// Summary renders a human-readable cluster summary.
func (c *Cluster) Summary() string {
	ctx := c.Scenario.Context
	return "Cluster: " + toString(ctx["cluster_id"]) + "\n" +
		"Region: " + toString(ctx["region"]) + "\n" +
		"Component: " + toString(ctx["component"]) + " v" + toString(ctx["version"]) + "\n" +
		"Problem: " + toString(firstNonEmpty(ctx["problem_desc"], ctx["problem"])) + "\n" +
		"Detail: " + toString(ctx["detail"])
}

func toString(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case int:
		return strconv.Itoa(t)
	case float64:
		return strconv.FormatFloat(t, 'f', -1, 64)
	default:
		return ""
	}
}

func firstNonEmpty(a, b any) any {
	if toString(a) != "" {
		return a
	}
	return b
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}
